pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedPrivateMethodRule.java

```diff
@@ -41,132 +39,159 @@
-                        .collect(Collectors.toSet());
-
-        Map<String, Set<ASTMethodDeclaration>> consideredNames =
-            file.descendants(ASTMethodDeclaration.class)
-                .crossFindBoundaries()
-                // get methods whose usages are all in this file
-                // TODO we could use getEffectiveVisibility here, but we need to consider overrides then.
-                .filter(it -> it.getVisibility() == Visibility.V_PRIVATE)
-                .filter(it -> !hasIgnoredAnnotation(it)
-                        && !hasExcludedName(it) 
-                        && !(it.getArity() == 0 && methodsUsedByAnnotations.contains(it.getName())))
-                .toStream()
-                .collect(Collectors.groupingBy(ASTMethodDeclaration::getName, HashMap::new, CollectionUtil.toMutableSet()));
+        visit(
+                file,
+                findPrivateMethods(file, methodsUsedByAnnotations(file)),
+                param);
+        return null;
+    }
 
-        file.descendants()
-            .crossFindBoundaries()
-            .map(NodeStream.<MethodUsage>asInstanceOf(ASTMethodCall.class, ASTMethodReference.class))
-            .forEach(ref -> {
-                String methodName = ref.getMethodName();
-                // the considered names might be mutated during the traversal
-                if (!consideredNames.containsKey(methodName)) {
-                    return;
-                }
-                JExecutableSymbol sym;
-                if (ref instanceof ASTMethodCall) {
-                    sym = ((ASTMethodCall) ref).getMethodType().getSymbol();
-                } else if (ref instanceof ASTMethodReference) {
-                    sym = ((ASTMethodReference) ref).getReferencedMethod().getSymbol();
-                } else {
-                    return;
-                }
-
-                JavaNode reffed = sym.tryGetNode();
-                if (reffed == null) {
-                    reffed = findDeclarationInCompilationUnit(file, sym);
-                }
-                if (reffed instanceof ASTMethodDeclaration
-                    && ref.ancestors(ASTMethodDeclaration.class).first() != reffed) {
-                    // remove from set, but only if it is called outside of itself
-                    Set<ASTMethodDeclaration> remainingUnused = consideredNames.get(methodName);
-                    if (remainingUnused != null
-                        && remainingUnused.remove(reffed) // note: side-effect
-                        && remainingUnused.isEmpty()) {
-                        consideredNames.remove(methodName); // clear this name
-                    }
-                }
-            });
-
-        // those that remain are unused
-        consideredNames.forEach((name, unused) -> {
+    private void visit(ASTCompilationUnit file,
+                       Map<String, Set<ASTMethodDeclaration>> privateMethods,
+                       Object param) {
+        filterUsedMethods(
```

Comment:
This should speak for itself: like [well-written prose](https://www.goodreads.com/quotes/7029841-clean-code-is-simple-and-direct-clean-code-reads-like#:~:text=Clean%20code%20reads%20like%20well,and%20straightforward%20lines%20of%20control.)



```
        filterUsedMethods().forEach().addViolation()
```